### PR TITLE
Deploy task

### PR DESCRIPTION
**Pull request by mpcarolin**
**Ticket**: [VIS-1102](https://jira.simpleviewtools.com/browse/vis-1102)

## Goal

* implement a deploy task in `tap-snack` that builds the app using EAS and submits it o Appetize

## Updates
* `appListServer`
  * a server for listing the submitted appetize apps 
* `tasks/definitions/appetize`
  * implemented a few more tasks for interacting with the appetize api
    * `remove`
    * `upsert`
* `tasks/definitions/deploy/deploy.js`
  * builds the app using eas, submits it to appetize, then outputs the results
* `tasks/utils/appetize/getParams.js`
  * adding `search`, `meta`, and `note` field mappings
* `tasks/utils/eas/findBuild.js`
  * helper for finding the latest EAS build given certain parameter filters  
* `tasks/utils/eas/findLatestBuild.js`
  * another helper that finds the latest build for a given platform, account, and expo project file path
* `tasks/utils/eas/getAccountName.js
  * helper for getting the current account name from the `eas` cli
* `tasks/utils/resolveTapRoot.js`
  * updated to use existing helper `getTapPath`
  * updated to work either *either* tap or location passed-in
  * added error handling
* `tasks/utils/sharedOptions.js`
  * added `file` and `url` options

## Testing
* `yarn test:tasks` -- ensure all tests pass
* ensure the github workflows succeeded with this PR: https://github.com/simpleviewinc/tap-visitapps-app/pull/6

### CODE REVIEW

File: appListServer/src/server.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+      res.status(status).json(apps)
+    }
+    catch (err) {
+      const status = get(err, 'response.status', 500)
+      const message = status.toString() === '401'
+        ? err.message + ' --- check your api token.'
+        : err.message
+
+      res.status(status).send(message)
```
Comment (lancetipton) [inline appListServer/src/server.js]: Since its a json api, we should respond with json

File: tasks/definitions/deploy/deploy.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+const { getPlatforms } = require(`${utilsPath}/getPlatforms`)
+const { resolveTapRoot } = require(`${utilsPath}/resolveTapRoot`)
+const { callApi } = require(`${utilsPath}/appetize/callApi`)
+const { eas } = require(`${utilsPath}/eas`)
+const fs = require('fs')
+
+/**
+ * Pushes a build to appetize
+ * @param {Object} options
+ * @param {Object} options.url
```
Comment (lancetipton) [inline tasks/definitions/deploy/deploy.js]: Can you define what the options are

File: tasks/package.json [hunk 3]
```diff
@@ -11,19 +11,23 @@
...
     "axios": "0.21.1",
     "dotenv": "9.0.2",
     "form-data": "4.0.0",
     "rimraf": "3.0.2"
   },
   "devDependencies": {
-    "@keg-hub/cli-utils": "0.2.0",
-    "@keg-hub/jsutils": "8.5.0"
+    "module-alias-jest": "0.0.3",
+    "@keg-hub/git-lib": "0.0.1",
```
Comment (lancetipton) [inline tasks/package.json:24]: why not make it a dependency , and not a dev dependency?

Then, you dont have to install it in other repos.

File: tasks/utils/resolveTapRoot.js [hunk 4]
```diff
@@ -1,24 +1,37 @@
 const path = require('path')
-const { snackRoot } = require('../paths')
-const { get } = require('@keg-hub/jsutils')
-const { getKegGlobalConfig, getAppRoot } = require('@keg-hub/cli-utils')
+const { getKegGlobalConfig, getTapPath } = require('@keg-hub/cli-utils')
+const { tryRequireSync } = require('@keg-hub/jsutils/src/node')
+
+/**
```
Comment (lancetipton) [inline tasks/utils/resolveTapRoot.js]: would be good as a helper in cli-utils

File: tasks/utils/appetize/getParams.js [hunk 5]
```diff
@@ -55,17 +56,19 @@ const getParams = args => {
     }
     case 'upsert': {
       addIfExists(url, params, 'url')
```
Comment (lancetipton) [inline tasks/utils/appetize/getParams.js:57]: we should probably make this a method that accepts a obj that accepts key value pairs and loops over them.
Then we can just call the method once, passing in the object

File: __mocks__/@keg-hub/git-lib.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+module.exports = {
```
Comment (lancetipton) [inline __mocks__/@keg-hub/git-lib.js]: This should be part of the git-lib repo, that way it does not have to be duplicated for every app the uses it. Then you can just import it directly when mocking the git-lib

**Review by lancetipton — CHANGES_REQUESTED** [review]
Looks good, just a couple items to look at

File: tasks/package.json (hunk 3 above)
Comment (mpcarolin) [reply tasks/package.json:24]: Agreed; updated

File: __mocks__/@keg-hub/git-lib.js (hunk 6 above)
Comment (mpcarolin) [reply __mocks__/@keg-hub/git-lib.js]: I actually determined I didn't need these mocks. `jest.mock` can already automock a package's whole api, so I updated this to just do that.

File: tasks/utils/appetize/getParams.js (hunk 5 above)
Comment (mpcarolin) [reply tasks/utils/appetize/getParams.js:57]: agreed. I updated this to do that.

File: tasks/utils/resolveTapRoot.js (hunk 4 above)
Comment (mpcarolin) [reply tasks/utils/resolveTapRoot.js]: updated

File: tasks/definitions/deploy/deploy.js (hunk 2 above)
Comment (mpcarolin) [reply tasks/definitions/deploy/deploy.js]: updated

File: appListServer/src/server.js (hunk 1 above)
Comment (mpcarolin) [reply appListServer/src/server.js]: updated

File: tasks/utils/appetize/getParams.js [hunk 7]
```diff
@@ -1,18 +1,47 @@
...
+/**
+ * 
+ * @param {Object} params - params, from the task, to be mapped
+ * @param {Array} acceptedKeys - valid param keys
+ * @returns 
  */
-const addIfExists = (item, params, key) => (
-  exists(item) && (params[key] = item)
-)
+const mapParamKeys = (params, acceptedKeys=[]) => {
```
Comment (lancetipton) [inline tasks/utils/appetize/getParams.js:33]: This is exactly what I was thinking. Well done. So much cleaner.

**Review by lancetipton — APPROVED** [review]
Looks good, thanks

**Pull request merged**